ui/src/app/workflows/components/workflow-logs-viewer/workflow-logs-viewer.tsx

```diff
@@ -34,6 +34,9 @@ export class WorkflowLogsViewer extends React.Component<WorkflowLogsViewerProps,
     public componentWillUnmount(): void {
         this.ensureUnsubscribed();
     }
+    private get error() {
+        return this.state.error && this.state.error.eventPhase !== Event.AT_TARGET && this.state.error;
```

Comment:
we need the AT_TARGET check to prevent error on simple disconnection

Reply:
The `eventPhase` is usually associated with the capturing/bubbling/at phase of an event in the DOM - since we're directly binding to the `onerror` property of the EventSource, when would this differ?

I'm actually wondering if maybe we should expose the status of the EventSource in the UI to give an indication of what the current state is (Closed, Connecting, Connected), then filter out the errors relating to the connection being closed.

Reply:
I did a bunch of digging on this today and think I can summarize what the underlying problem here is.  Basically, when an EventSource is unexpectedly closed, an error event is emitted and the browser will attempt to reconnect the EventSource - those are the errors that you're seeing.

Those errors are also very common here since when the API finishes streaming logs from a completed pod, it closes the connection, however the browser is not aware that it's the reason it's being closed.  Now the reason we don't see the re-connection happen immediately is that when an error is emitted in RxJS, it's hard-coded to unsubscribe from the stream (https://iamturns.com/continue-rxjs-streams-when-errors-occur/), however that's also the reason you'll immediately see an unnecessary request to the artifact if available since that's the fallback code.

There still an actual bug here though - if the pod logs aren't finished streaming, *and* the pod is not yet finished running, then we will not continue streaming logs (like if you killed your wifi).

The actual fix to this would be when an error occurs, check to see whether or not the node that you're currently streaming logs for is still running or pending - if it is, re-connect the EventSource, else allow it to die gracefully.

I have some code that I cooked up to do exactly that which I can push to a branch if you want to take a look.  I also re-ordered the code to look at the artifacts first since that gracefully handles the case where the Pod was deleted, although there's an interesting problem when running the server locally with minio since your local server generally can't resolve a `.cluster.local` address.  It still works, it just takes a bit since you need to wait for the request to timeout.

Reply:
@dcherman I'd like to see your work. I think you may have a better handle than I do for this. Would you like to take ownership? One thought - if this is merged as a short-term fix - you can do something longer term and more correct?

Reply:
Sure, happy to take ownership.  I just pushed up the code to https://github.com/dcherman/argo/commit/e0fb94daad40c51b60cb6747636f068342ec6e45.  There's a couple things left to handle there (error handling when offline).  The biggest thing I would want input on are:

* Any opposition to introducing Jest as a test framework here?  Since this doesn't really touch the React part of the code, it should be a pretty minimal setup.  We can introduce support for react-enzyme and things as needed if/when we start testing that code.

* When using an artifact repo that isn't externally available from the cluster (say, minio that's not exposed), running the UI/server locally is not a great user experience since it takes ~30s for the request to fail/timeout since you generally can't resolve `.cluster.local` addresses.  Would want to figure out how we can handle that a little better.

https://github.com/dcherman/argo/commit/e0fb94daad40c51b60cb6747636f068342ec6e45

Reply:
Thank you @dcherman. @simster7 is working on on a test framework, and maybe able to provide details.

I'd like to aim for simpler code around this, as it has server bugs from just being complex.

Reply:
Footnote - this is why I removed the fallback to artifacts. Buggy and not needed.

Reply:
Since I believe the Workflow Archive uses the same code to view logs, it's actually really nice to continue loading logs from the artifact repo since for archived workflows, it's likely that the actual Pod has already been deleted.

If we want to maintain that and still simplify the code, what would you think about having the "live" workflow UI read logs from the cluster as you're doing here, but have the archive view only read from the artifacts repo?

Essentially remove the fallback code entirely, but have something like `archived ? getLogsFromArtifact : getLogsFromCluster`

Reply:
not sure about that - it is possible that you have the workflow archive enabled - but you don't have log archiving enabled - at that point you want be able to see any pod logs 

however - if you think you can reliably fix the fallback - then that would be fine too

Reply:
Actually incidentally, the code I have already fixes that :).  In the event that log archiving is not enabled, it detects that no logs can be loaded from the artifact repo and immediately throws an error; it doesn't even try to load the logs.  While it's true that you would not be able to see logs since at that point we have nowhere to load them from, you would also see the error message in the UI saying `no artifact logs are available`

Reply:
when will you raise the PR please?

Reply:
https://github.com/argoproj/argo/pull/4150